config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/TrustedCertEntityBuilder.java

```diff
@@ -0,0 +1,138 @@
+        ).collect(Collectors.toList());
+    }
+
+    private Entity buildTrustedCertEntity(String name, TrustedCert trustedCert, Map<String, String> certificateFiles) {
+        final String id = idGenerator.generate();
+        final Element trustedCertElem = createElementWithAttributesAndChildren(
+                document,
+                TRUSTED_CERT,
+                ImmutableMap.of(ATTRIBUTE_ID, id),
+                createElementWithTextContent(document, NAME, name),
+                buildCertData(trustedCert, certificateFiles)
+        );
+        buildAndAppendPropertiesElement(trustedCert.getProperties(), document, trustedCertElem);
+
+        return new Entity(TRUSTED_CERT_TYPE, name, id, trustedCertElem);
+
+    }
+
+    private Element buildCertData(TrustedCert trustedCert, Map<String, String> certificateFiles) {
+        if (StringUtils.isNotEmpty(trustedCert.getUrl())) {
+            return buildCertDataFromUrl(trustedCert);
+        } else if (StringUtils.isNotEmpty(trustedCert.getFile())) {
+            return buildCertDataFromFile(certificateFiles.get(trustedCert.getFile()));
+        } else if (trustedCert.getCertificateData() != null) {
+            CertificateData certData = trustedCert.getCertificateData();
+            return createElementWithChildren(
+                    document,
+                    CERT_DATA,
+                    createElementWithTextContent(document, ISSUER_NAME, certData.getIssuerName()),
+                    createElementWithTextContent(document, SERIAL_NUMBER, certData.getSerialNumber()),
+                    createElementWithTextContent(document, SUBJECT_NAME, certData.getSubjectName()),
+                    createElementWithTextContent(document, ENCODED, certData.getEncodedData())
+            );
+        } else {
+            throw new EntityBuilderException("Trusted Cert must be loaded from a valid url, from a file, or using certificateData reference.");
+        }
+    }
+
+    private Element buildCertDataFromFile(String certFileLocation) {
+        if (StringUtils.isEmpty(certFileLocation)) {
+            throw new EntityBuilderException("The certificate file location specified does not exist.");
+        }
+
+        try (FileInputStream is = new FileInputStream(certFileLocation)) {
+            CertificateFactory certFact = CertificateFactory.getInstance("X.509");
+            X509Certificate cert = (X509Certificate) certFact.generateCertificate(is);
+            return createCertDataElementFromCert(cert);
+        } catch (IOException e) {
+            throw new EntityBuilderException("The certificate file location specified does not exist.");
+        } catch (CertificateException e) {
+            throw new EntityBuilderException("Error generating certificate: ", e);
+        }
+    }
+
+    private Element buildCertDataFromUrl(TrustedCert trustedCert) {
+        HttpsURLConnection conn = null;
+        try {
```

Comment:
I tried doing that, but I think HttpsUrlConnection doesn't implement AutoClosable. Also should I retrieve certs this way or should I use a different library? I'm not too sure if this would work with an ldap?

Reply:
I think it's better keep using only java default things if it's working and not too much code.

For ldap you may need to test it, try checking [openldap](https://www.openldap.org/) and starting it with certs enabled, if not too hard.